network_plugin/__init__.py

```diff
@@ -237,9 +237,15 @@ def get_ondemand_public_ip(vca_client, gateway, ctx):
         try to allocate new public ip for ondemand service
     """
     old_public_ips = set(gateway.get_public_ips())
+    allocated_ips = set([address.external
+                         for address in collectAssignedIps(gateway)])
+    available_ips = old_public_ips - allocated_ips
+    if available_ips:
+        new_ip = list(available_ips)[0]
+        ctx.logger.info("Public IP {0} was reused.".format(new_ip))
+        return new_ip
     for i in range(5):
         ctx.logger.info("Try to allocate public IP")
-        wait_for_gateway(vca_client, gateway.get_name(), ctx)
```

Comment:
Looks as it will be regression in case several thread that tried to change Nat rules